src/oc_erchef/schema/deploy/create_and_update_users.sql

```diff
@@ -0,0 +1,92 @@
+                                    p_admin users.admin%TYPE) RETURNS void AS $$
+  BEGIN
+    INSERT INTO users
+      (id, authz_id, username, email, public_key, hashed_password, salt, hash_type,
+       last_updated_by, created_at, updated_at, external_authentication_uid,
+       recovery_authentication_enabled, serialized_object, admin,
+       pubkey_version)
+    VALUES (p_id, p_authz_id, p_username, p_email, 'thisisnotakey',
+       p_hashed_password, p_salt, p_hash_type, p_last_updated_by, p_created_at, p_updated_at,
+       p_external_authentication_uid, p_recovery_authentication_enabled,
+       p_serialized_object, p_admin, p_pubkey_version);
+    IF p_public_key != 'thisisnotakey' THEN
+      INSERT INTO keys
+        (id, key_name, public_key, key_version, created_at,  expires_at)
+      VALUES
+         (p_id, 'default', p_public_key, p_pubkey_version, now(), 'infinity'::timestamp);
+    END IF;
+  END
+$$ LANGUAGE plpgsql;
+
+DROP FUNCTION IF EXISTS update_user(integer,text,text,text,password_hash_type,text,text,boolean,text,text,character,timestamp without time zone,boolean,character);
+
+CREATE OR REPLACE FUNCTION update_user(p_pubkey_version users.pubkey_version%TYPE,
+                                       p_public_key users.public_key%TYPE,
+                                       p_hashed_password users.hashed_password%TYPE,
+                                       p_salt users.salt%TYPE,
+                                       p_hash_type users.hash_type%TYPE,
+                                       p_serialized_object users.serialized_object%TYPE,
+                                       p_external_authentication_uid users.external_authentication_uid%TYPE,
+                                       p_recovery_authentication_enabled users.recovery_authentication_enabled%TYPE,
+                                       p_email users.email%TYPE,
+                                       p_username users.username%TYPE,
+                                       p_last_updated_by users.last_updated_by%TYPE,
+                                       p_updated_at users.updated_at%TYPE,
+                                       p_admin users.admin%TYPE,
+                                       p_id character(32)) RETURNS void AS $update_user$
+    BEGIN
+    IF p_public_key != 'thisisnotakey' THEN
+      IF p_public_key IS NOT NULL THEN
+        UPDATE keys SET public_key = p_public_key,
+                        key_version = p_pubkey_version,
+                        expires_at = 'infinity'::timestamp
+        WHERE id = p_id AND key_name = 'default';
+       INSERT INTO keys (id, key_name, public_key, key_version, created_at, expires_at)
+         SELECT p_id, 'default', p_public_key, p_pubkey_version, now(), 'infinity'::timestamp
+         WHERE NOT EXISTS (SELECT 1 FROM keys WHERE id = p_id AND key_name = 'default');
+      ELSE
+       DELETE FROM keys WHERE id = p_id AND key_name = 'default';
+      END IF;
+    END IF;
```

Comment:
we still need this because we will not be inserting the key into the users table in all the new requests (after this release) so just leaving the trigger in will update 'thisisnotakey' into the keys table. This function does the right things and the trigger will only be useful for older versions of chef-server

Reply:
In SQL NULL returns NULL for all comparisons. So the DELETE will never fire above I think, but instead NULL != to 'thisisnotakey' will return NULL, which is falsy, and the inner if will never be reached.

```
CREATE OR REPLACE FUNCTION test(val text)
RETURNS text
AS $$
DECLARE
  result text;
BEGIN
   IF val != 'A' THEN
     IF val IS NOT NULL THEN
      result := 'NOT A, NOT NULL';
     ELSE 
      result := 'NOT A, BUT NULL';
     END IF;
   ELSE
      result := 'A';
   END IF;
   RETURN result;
END;
$$;
```

```
opscode_chef=# select test(NULL);
 test 
------
 A
```

Reply:
We probably should have unit test coverage of some of these functions to insure they work as expected.

Reply:
That is an awesome catch! updated the cases to fall through to the delete operation when expected.

I will check the unit tests in bifrost and add something to test this functionality.